src/AlarmClock.h

```diff
@@ -56,87 +72,71 @@ template<typename Duration> class AlarmClock {
 
 protected:
 
-   void AlarmClockThread() {
-      SleepTimeIsBelow500ms() ? AlarmClock::SleepForFullAmount() : AlarmClock::SleepInIntervals();
+   void AlarmClockInterruptableThread() {
+      do {
+         mSleepFunction(kSleepTimeUsCount);
+         // This is required because the unit tests use their own "fake sleep"
+         // pass it into the constructor. By doing that, the unit tests will
+         // fail if the mExpired.store(true) is in the sleep function instead
+         // of the thread function. aka Due to unit test limitation.
+         if (!mExit && !mReset) {
+            mExpired.store(true);
+         }
+         if (mExit) {  
+            break;
+         }
+
+         if (mExpired) {
+            while (!mReset && !mExit) {
+               this_thread::sleep_for(chrono::microseconds(1));
+            }
+         }
+         mReset.store(false);
+      } while (!mExit);
    }
   
    void StopBackgroundThread() {
-      mExpired.store(true);
-      mExited.wait();
-   }
-
-   bool SleepTimeIsBelow500ms() {
-      return Duration(kSleepTime) <= microseconds(kSmallestIntervalInUS);
-   }
-
-   void SleepForFullAmount() {
-      mSleepFunction(kSleepTimeUsCount);
-      mExpired.store(true);
+      if (mTimerThread.joinable() && mExit) {
+         mTimerThread.join();
+      }
    }
 
    void SleepUs(unsigned int t) {
-      std::this_thread::sleep_for(microseconds(t));
-   }
-   
-   // If 500ms is NOT an even divisor of the amount of sleep
-   //    time given, should sleep for every multiple of 500ms
-   //    that divides, then check the remaining time and just
-   //    sleep for only that last bit.
-   //
-   // If 500ms IS an even divisor of the sleep time, sleep for 
-   //    all but one multiple of 500ms. There is overhead induced
-   //    by the while loop, so if the full amount of time is slept,
-   //    the precision of the alarm clock will suffer.
-   size_t GetNumberOfSleepIntervals() {
-      return (kSleepTimeMsCount % kSmallestIntervalInMS == 0) ?
-             ((kSleepTimeMsCount/kSmallestIntervalInMS) - 1) :
-             (kSleepTimeMsCount/kSmallestIntervalInMS);
-   }
-  
-   void SleepForRemainder(const unsigned int& currentSleptFor) {
-      if (currentSleptFor < kSleepTimeUsCount) {
-         mSleepFunction(kSleepTimeUsCount - currentSleptFor);
-      }
-   } 
-
-   void SleepInIntervals() {
-      StopWatch timer;
-      size_t numberOfSleeps = GetNumberOfSleepIntervals();
-      while (KeepRunning() && numberOfSleeps > 0) {
-         mSleepFunction(kSmallestIntervalInUS);
-         --numberOfSleeps; 
-      }
-      auto currentSleptFor = timer.ElapsedUs();
-      if (KeepRunning()) {
-         SleepForRemainder(currentSleptFor);
+      StopWatch sw;
```

Comment:
@KjellKod We found that the check of the two atomics (mExit and mReset) were causing an overhead of 50 ms. Using relaxed atomics did not improve that time so we decided to keep the Stopwatch implementation. I'll have a PR up later today (post lunch) with the changes.